File: deploy/application/deploy_service.go
package application

import (
	"fmt"
	"time"

	"github.com/koyeo/nest/deploy/domain"
	"github.com/koyeo/nest/i18n"
)

// DeployService orchestrates the deploy conflict resolution flow.
type DeployService struct {
	fs       domain.RemoteFS
	exec     domain.RemoteExec
	snapshot domain.SnapshotRepository
	prompter domain.UserPrompter
	lang     string
}

// NewDeployService creates a new DeployService with all dependencies injected.
func NewDeployService(
	fs domain.RemoteFS,
	exec domain.RemoteExec,
	snapshot domain.SnapshotRepository,
	prompter domain.UserPrompter,
	lang string,
) *DeployService {
	return &DeployService{
		fs:       fs,
		exec:     exec,
		snapshot: snapshot,
		prompter: prompter,
		lang:     lang,
	}
}

// Deploy handles extracting, conflict resolution, file deployment, and snapshot update.
//
// Parameters:
//   - bundleRemotePath: path to the uploaded tar.gz on the remote server
//   - targetDir: the target directory where files should be deployed
//   - bundleName: name of the bundle (e.g., "app.tar.gz")
//   - bundleHash: SHA256 hash of the bundle
func (s *DeployService) Deploy(bundleRemotePath, targetDir, bundleName, bundleHash, conflictStrategy string) error {
	nestDir := fmt.Sprintf("%s/.nest", targetDir)
	tmpDir := fmt.Sprintf("%s/.nest/tmp", targetDir)

	// Ensure .nest directory exists
	_ = s.fs.MkdirAll(nestDir)

	// Extract to .nest/tmp/
	cmd := fmt.Sprintf("rm -rf %s && mkdir -p %s && tar -xzf %s -C %s", tmpDir, tmpDir, bundleRemotePath, tmpDir)
	if err := s.exec.Exec(cmd); err != nil {
		return fmt.Errorf("extract bundle error: %s", err)
	}
	defer func() {
		// Clean up .nest/tmp/
		_ = s.exec.Exec(fmt.Sprintf("rm -rf %s", tmpDir))
	}()

	// List extracted files
	extractedFiles, err := s.fs.ReadDir(tmpDir)
	if err != nil {
		return fmt.Errorf("list extracted files error: %s", err)
	}

	// Detect conflicts
	var conflictFiles []string
	for _, f := range extractedFiles {
		filePath := fmt.Sprintf("%s/%s", targetDir, f)
		if _, statErr := s.fs.Stat(filePath); statErr == nil {
			conflictFiles = append(conflictFiles, f)
		}
	}

	// Read existing snapshot
	snap, err := s.snapshot.Read(targetDir)
	if err != nil {
		return fmt.Errorf("read snapshot error: %s", err)
	}

	// Resolve conflicts
	if len(conflictFiles) > 0 {
		if err = s.resolveConflicts(targetDir, conflictFiles, snap, conflictStrategy); err != nil {
			return err
		}
	}

	// Move files from .nest/tmp/ to target directory
	for _, f := range extractedFiles {
		src := fmt.Sprintf("%s/%s", tmpDir, f)
		dst := fmt.Sprintf("%s/%s", targetDir, f)
		cmd = fmt.Sprintf("mv %s %s", src, dst)
		if err = s.exec.Exec(cmd); err != nil {
			return fmt.Errorf("move file error: %s", err)
		}
	}

	// Build file records for snapshot
	var fileRecords []domain.FileRecord
	for _, f := range extractedFiles {
		filePath := fmt.Sprintf("%s/%s", targetDir, f)
		hash, _ := s.fs.FileHash(filePath)
		modTime, mtErr := s.fs.FileModTime(filePath)
		if mtErr != nil {
			modTime = time.Now().UTC()
		}
		fileRecords = append(fileRecords, domain.FileRecord{
			Path:    f,
			Hash:    hash,
			ModTime: modTime,
		})
	}

	// Update snapshot
	entry := domain.NewSnapshotEntry(bundleName, bundleHash, fileRecords)
	isNew := snap == nil
	if isNew {
		snap = &domain.Snapshot{}
	}
	snap.AddEntry(entry)

	if err = s.snapshot.Write(targetDir, snap); err != nil {
		return fmt.Errorf("write snapshot error: %s", err)
	}

	fmt.Println(i18n.Msg(i18n.MsgDeployComplete, s.lang))
	return nil
}

// resolveConflicts handles managed and unmanaged file conflicts.
func (s *DeployService) resolveConflicts(targetDir string, conflictFiles []string, snap *domain.Snapshot, conflictStrategy string) error {
	result := domain.ClassifyConflicts(conflictFiles, snap)

	// Remove Nest-managed files silently
	for _, f := range result.ManagedFiles {
		if err := s.fs.Remove(fmt.Sprintf("%s/%s", targetDir, f)); err != nil {
			return fmt.Errorf("remove managed file error: %s", err)
		}
	}

	// Handle unmanaged files
	if len(result.UnmanagedFiles) > 0 {
		var action domain.ConflictAction
		var suffix string
		var err error

		switch conflictStrategy {
		case "overwrite":
			action = domain.ActionOverwrite
		case "backup":
			action = domain.ActionBackup
			suffix = ".bak"
		case "error":
			return fmt.Errorf("unmanaged file conflicts: %v", result.UnmanagedFiles)
		default:
			// Interactive mode
			action, suffix, err = s.prompter.AskConflictAction(result.UnmanagedFiles, s.lang)
			if err != nil {
				return err
			}
		}

		for _, f := range result.UnmanagedFiles {
			filePath := fmt.Sprintf("%s/%s", targetDir, f)
			switch action {
			case domain.ActionBackup:
				backupName := domain.NextBackupName(f, suffix, func(candidate string) bool {
					_, statErr := s.fs.Stat(fmt.Sprintf("%s/%s", targetDir, candidate))
					return statErr == nil
				})
				fmt.Println(i18n.Msgf(i18n.MsgBackingUp, s.lang, f, backupName))
				if err = s.fs.Rename(filePath, fmt.Sprintf("%s/%s", targetDir, backupName)); err != nil {
					return fmt.Errorf("backup file error: %s", err)
				}
			case domain.ActionRemove, domain.ActionOverwrite:
				if action == domain.ActionRemove {
					fmt.Println(i18n.Msgf(i18n.MsgRemoving, s.lang, f))
				}
				if err = s.fs.Remove(filePath); err != nil {
					return fmt.Errorf("remove file error: %s", err)
				}
			}
		}
	}

	return nil
}

File: apps/ship/src/deploy/application/deploy-service.ts
import { REMOTE_META_DIR } from "../../common/const.js";
import { nextBackupName } from "../domain/backup.js";
import { classifyConflicts } from "../domain/conflict.js";
import type { RemoteExec, RemoteFS, SnapshotRepository, UserPrompter } from "../domain/interfaces.js";
import { type FileRecord, type Snapshot, addEntry, emptySnapshot, newSnapshotEntry } from "../domain/snapshot.js";

export interface DeployBundle {
  /** Path of the uploaded tar.gz on the remote server. */
  bundleRemotePath: string;
  targetDir: string;
  bundleName: string;
  /** SHA256 of the bundle. */
  bundleHash: string;
}

/** Extract → conflict resolution → move → snapshot update. Both SFTP and cloud-storage paths converge here. */
export class DeployService {
  constructor(
    private readonly fs: RemoteFS,
    private readonly exec: RemoteExec,
    private readonly snapshots: SnapshotRepository,
    private readonly prompter: UserPrompter,
    private readonly now: () => Date,
  ) {}

  async deploy(bundle: DeployBundle): Promise<void> {
    const { targetDir } = bundle;
    const metaDir = `${targetDir}/${REMOTE_META_DIR}`;
    const tmpDir = `${metaDir}/tmp`;

    await this.fs.mkdirAll(metaDir);
    await this.exec.exec(`rm -rf ${tmpDir} && mkdir -p ${tmpDir} && tar -xzf ${bundle.bundleRemotePath} -C ${tmpDir}`);
    try {
      const extracted = await this.fs.readDir(tmpDir);

      const conflicts: string[] = [];
      for (const f of extracted) {
        if ((await this.fs.stat(`${targetDir}/${f}`)) !== null) {
          conflicts.push(f);
        }
      }

      const existing = await this.snapshots.read(targetDir);
      if (conflicts.length > 0) {
        await this.resolveConflicts(targetDir, conflicts, existing);
      }

      for (const f of extracted) {
        await this.exec.exec(`mv ${tmpDir}/${f} ${targetDir}/${f}`);
      }

      const records: FileRecord[] = [];
      for (const f of extracted) {
        const path = `${targetDir}/${f}`;
        // Directories have no content hash; nest records "" for them (hash, _ := FileHash).
        const info = await this.fs.stat(path);
        const hash = info !== null && !info.isDir ? await this.fs.fileHash(path) : "";
        records.push({ path: f, hash, mod_time: (await this.fs.fileModTime(path)).toISOString() });
      }

      const base: Snapshot = existing ?? emptySnapshot();
      const next = addEntry(base, newSnapshotEntry(bundle.bundleName, bundle.bundleHash, records, this.now()));
      await this.snapshots.write(targetDir, next);
      process.stdout.write("  ✅ Deploy complete\n");
    } finally {
      await this.exec.exec(`rm -rf ${tmpDir}`);
    }
  }

  private async resolveConflicts(targetDir: string, conflicts: string[], snapshot: Snapshot | null): Promise<void> {
    const result = classifyConflicts(conflicts, snapshot);

    for (const f of result.managedFiles) {
      await this.fs.remove(`${targetDir}/${f}`);
    }

    if (result.unmanagedFiles.length === 0) {
      return;
    }
    const decision = await this.prompter.askConflictAction(result.unmanagedFiles);
    for (const f of result.unmanagedFiles) {
      const path = `${targetDir}/${f}`;
      if (decision.action === "backup") {
        const siblings = new Set(await this.fs.readDir(targetDir));
        const backupName = nextBackupName(f, decision.suffix, (c) => siblings.has(c));
        process.stdout.write(`  📦 Backup: ${f} → ${backupName}\n`);
        await this.fs.rename(path, `${targetDir}/${backupName}`);
      } else {
        process.stdout.write(`  🗑  Remove: ${f}\n`);
        await this.fs.remove(path);
      }
    }
  }
}
